pkg/mimirpb/prealloc_rw2.go

```diff
@@ -30,27 +34,50 @@ func FromWriteRequestToRW2Request(rw1 *WriteRequest, commonSymbols []string, off
 	defer reuseSymbolsTable(symbols)
 	symbols.ConfigureCommonSymbols(offset, commonSymbols)
 
+	var metadataMap map[string]*MetricMetadata
+	if len(rw1.Metadata) > 0 {
+		metadataMap = make(map[string]*MetricMetadata, len(rw1.Metadata))
+		for _, meta := range rw1.Metadata {
+			metadataMap[meta.MetricFamilyName] = meta
+		}
+	}
+
 	rw2Timeseries := make([]TimeSeriesRW2, 0, len(rw1.Timeseries)+len(rw1.Metadata)) // TODO: Pool-ify this allocation
 	for _, ts := range rw1.Timeseries {
 		refs := make([]uint32, 0, len(ts.Labels)*2) // TODO: Pool-ify this allocation
+		metricName := ""
 		for i := range ts.Labels {
+			if ts.Labels[i].Name == labels.MetricName {
+				metricName = ts.Labels[i].Value
+			}
 			refs = append(refs, symbols.Symbolize(ts.Labels[i].Name), symbols.Symbolize(ts.Labels[i].Value))
 		}
 
+		var metadata MetadataRW2
+		if len(rw1.Metadata) > 0 {
```

Comment:
[nit] Isn't more accurate checking `metadataMap` instead?
```suggestion
		if len(metadataMap) > 0 {
```

Reply:
Fair enough, anything that guards when the map doesn't exist. Checking the map itself is better if we change the conditions for initializing it